### PR TITLE
JAMES-1801 Convert JMAP filter to SearchQuery 

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/FilterToSearchQuery.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    public SearchQuery convert(Filter filter) {
+        if (filter instanceof FilterCondition) {
+            return convertCondition((FilterCondition) filter);
+        }
+        if (filter instanceof FilterOperator) {
+            SearchQuery searchQuery = new SearchQuery();
+            searchQuery.andCriteria(convertOperator((FilterOperator) filter));
+            return searchQuery;
+        }
+        throw new RuntimeException("Unknown filter");
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/FilterToSearchQuery.java]: can you add the filter class in the error message?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/FilterConditionTest.java [hunk 2]
```diff
@@ -69,101 +69,50 @@ public void builderShouldBuildWhenGivenNotInMailboxesAsEllipsis() {
...
-        FilterCondition.builder().header(ImmutableList.of());
-    }
 
     @Test
     public void buildShouldWork() {
-        FilterCondition expectedFilterCondition = new FilterCondition(Optional.of(ImmutableList.of("1")), Optional.of(ImmutableList.of("2")), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), 
-                Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), 
-                Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty());
+        Date before = new Date();
+        Date after = new Date();
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/FilterConditionTest.java]: can you fix dates?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+                            .build()),
+                    FilterCondition.builder()
+                        .cc(cc)
+                        .build()
+                ));
+
+        SearchQuery searchQuery = new FilterToSearchQuery().convert(complexFilter);
+
+        assertThat(searchQuery).isEqualTo(expectedSearchQuery);
+    }
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java]: Can you add a test with an empty filterCondition?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/FilterToSearchQuery.java (hunk 1 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/FilterToSearchQuery.java]: OK

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/FilterConditionTest.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/FilterConditionTest.java]: OK

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java]: OK

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java [hunk 4]
```diff
@@ -41,6 +42,25 @@
 public class FilterToSearchQueryTest {
 
     @Test
+    public void filterConditionShouldThrowWhenUnknownFilter() {
+        assertThatThrownBy(() -> new FilterToSearchQuery().convert(new Filter() {
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java]: you can write `Filter myFilter = (indentation -> null);`

Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java]: Right

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+        String to = "recipient@james.org";
+        String subject = "subject";
+        SearchQuery expectedSearchQuery = new SearchQuery();
+        expectedSearchQuery.andCriteria(SearchQuery.and(ImmutableList.of(
+                SearchQuery.address(AddressType.From, from),
+                SearchQuery.address(AddressType.To, to),
+                SearchQuery.headerContains("Subject", subject))));
+
+        Filter complexFilter = FilterOperator.and(
+                FilterCondition.builder()
```
Comment (mbaechler) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java]: The jmap query is really mapped to this ? not a more complex FilterCondition ?

Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java]: OK

**Comment by rouazana** [comment]
:+1:

**Comment by mbaechler** [comment]
👍
